cilium: encryption, fixes for ENI & Azure mode with shared podIPs and networkIPs

jrfastab opened the pull request:
Subnet mode came with some baked in assumptions that are not met in all ENI use cases. First we assumed the networkIPs would not overlap with the podIPs. This was important because we use the same routing table for both encrypt and decrypt with the above assumption this works because encrypted packets and decrypted packets will match different routes and never collide. But, once this assumption is broke depending on how subnets and IPs are chosen its possible they collide. The result is a encrypted packet may be routed as if it has already been decrypted or vice versa. We've seen some cases where the overlap happens in EKS and AKS.

Next we made an assumption that cilium_host interface would always understand how to handle packets. In ENI mode with endpoint routes enabled we optimized cilium_host recently so it would not handle ingress packets. However, we missed the subnet encryption case combination of features. Fix it here by also enabling subnet encryption to work with endpoint routes. We took this opportunity to make a small optimization and cut out an extra pass through bpf_network as it was no longer needed with above fixes.

Finally, feature detection incorrectly detected the ingress interface in some cases.

Quick note, I broke the patches down into small chunks. Other than the refactor patches they are hopefully small and straightforward. Patch List 

 patches 1- 2 refactor code so we can push optional flag down to policy rules and program fwd policy directly
 patch3: set optional flag on all fwd policies, we don't use fwd polices for policy (to drop packets not encrypted for example) because we can enforce these policies in the BPF programs and drop any non-compliant packets early in the stack at XDP or TC layers. Without this endpoint routes + one CIDR for network and podIPs can cause a drop.
 patch4: subnetEncryption was always enabling nodeEncrypt even if it wasn't configured this is wrong.
 patch5: ENI needs to use linkAddr for IPSec tunnel IPs
 patch6: ENI, redirecting to cilium_host breaks endpoint routes
 patch7: simplify stack to avoid extra loop by having xfrm stack clear mark bits.
 patch8: remove mark from fwd rule to make it more permissive again we don't use it for policy
 patch9: clear mark from xfrm states so we can skip extra pass through eth0
 patch10: ENI auto iface detection doesn't in my deployment it picks the wrong interface, this makes it so we use the user supplied interface if its given.

Couple pictures to help with understanding ingress datapath.
 
Picture of possible failing case if subnet IPs overlap and then collide with interface IPs.

![encryption notes(2)](https://user-images.githubusercontent.com/1247472/108762580-1f93d380-7505-11eb-8f20-689e2c976098.png) 


Picture of fixed datapath to support overlapping subnet IPs.

![encryption notes(1)](https://user-images.githubusercontent.com/1247472/108762569-1c004c80-7505-11eb-9a4c-d459cc67d858.png)

jrfastab committed:
cilium: encryption, allow programming IPSec FWD policy and template IPs

So far we've programmed FWD policy rules to mirror the IN policy rules,
but this requires the outer IPSec tunnel dstIP and inner dstIP cidr to
be non-overlapping. Also we've been setting the template IPs the same
as the matching src/dst IPs. We are about to fix the ENI case where these
assumptions are no longer true.

In order to support this allow pushing unique FWD/IN policies and
also specify matching IPs and template IPs.

Note: It probably makes sense to do a heavier refactor here and rip
out UpsertIPsecEndpoint(), at this point it just adds confusion in
my opinion. But, lets avoid that at the moment for backports.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, set optional flag for fwd policy rules

The XFRM stack will check for fwd policies after a decrypt or encrypt
operation is done. We don't really need these for policy reasons
because Cilium BPF side is enforcing policy. We can't just delete
the rules though. If we did the decrypted traffic, which will still
have the xfrm context encoded in the skb, will fail policy match and
generate a NOPOLICY error. But as is in ENI modes the template lookup
is failing with an TMPL_MISMATCH error.

To fix make the fwd rules less strict by adding the optional bit
to the rule. This will skip the template lookup altogether on the
kernel side.

Although this is needed for ENI cases where the packet is handled
by the routing stack we can enable it in all cases. Making other
cases less strict is also OK.

Before this we had fwd `ip x p` like this,

root@ip-192-168-90-141:/home/cilium# ip x p
src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.0.0
		proto esp reqid 1 mode tunnel

Now we have a 'level use' included indicating the policy is not strict,

root@ip-192-168-90-141:/home/cilium# ip x p
src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.0.0
		proto esp reqid 1 mode tunnel
		level use

This corresponds to kernel code,

 static inline int
 xfrm_policy_ok(const struct xfrm_tmpl *tmpl, const struct sec_path *sp, int start,
	       unsigned short family)
 {
	int idx = start;

	if (tmpl->optional) {
		if (tmpl->mode == XFRM_MODE_TRANSPORT)
			return start;
	} else
		start = -1;
	for (; idx < sp->len; idx++) {
		if (xfrm_state_ok(tmpl, sp->xvec[idx], family))
			return ++idx;
		if (sp->xvec[idx]->props.mode != XFRM_MODE_TRANSPORT) {
			if (start == -1)
				start = -2-idx;
			break;
		}
	}
	return start;
 }

And with 'level use' we hit the first 'if (tmpl->optional)' so that the
subsequent xfrm_state_ok checks are skipped.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, ENI with subnet encryption needs rules for link IP

In ENI mode we use link IP for outer IPsec headers. In order for rules
to be configured with correct IPs do a link lookup to discover the
correct IP.

Without this patch we will be missing the 'in' policy rules shown here,

 root@ip-192-168-90-141:/home/cilium# ip x p
 src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.90.141
		proto esp reqid 1 mode tunnel
		level use
 src 0.0.0.0/0 dst 192.168.64.0/19
 	dir in priority 0 ptype main
	mark 0xd00/0xf00
	tmpl src 0.0.0.0 dst 192.168.90.141

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, fix redirect when endpoint routes enabled

If we have endpoint routes enabled we want to use the routing table
to find the lxc* device. Trying to do this with redirects doesn't
work and may cause cilium_host to drop the packet.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, ENI mode uses overlapping CIDR for outer/inner IPs

TLDR: Remove 0xd00 rule/routes and let xfrm stack clear mark, but restrict
      overlapping pod/network CIDRs to 4.19 and above kernels.

Because ENI mode uses an outer IP that overlaps with the inner pod IPs we
need to be extra careful with the routing/rule configurations. For example
we currently put a route in the decrypt routing table that looks like
this,

	192.168.0.0/16 dev cilium_host mtu 9001

where 192.168.0.0/16 is the subnet IP in this example. The purpose of this
rule is twofold. First in the encryption direction it should match the
encrypted packet and send it to cilium_host for necessary rewrites. Second
on decryption there should be a simliar rule to match and send traffic to
the sending interface. Something like this,

	$POD_CIDR dev eth0 mtu

But, in the ENI case the decrypted CIDR and the encrypted CIDR are the
same so they collide and we get a single rule. Worse now received ESP
packets will match above and be routed to cilium_host before they can
be decrypted. cilium_host will become confused and send them to the
stack. At this point the mark set by bpf_network will be dropped and
the stack will miss the xfrm state lookup.

  eth0 (rx esp) -> bpf_packet (mark for decryption mark=0xd00)
                     -> rule finds 0x0d00 uses routing table 200
		     -> route hits {$CIDR dev cilium_host}
		     -> ESP packet on cilium host runs bpf_host
		     -> bpf_host return CTX_ACT_OK
		     -> cilium_host fwd to cilium_net mark = 0 now
		     -> ESP packet (dev=cilium_net) lookup xfrm state
		     -> XFRM_NOSTATE error because mark cleared

So everything went sideways after the 0xd00 routing table hit. The reason
we even have a 0xd00 routing rule is to clear the mark bits after decryption,
but before pushing to stack. Expected flow,

   eth0 (rx esp) -> bpf_packet (mark for decryption mark=0xd00)
                     -> rule finds 0x0d00 uses routing table 200
		     -> no match send to stack with mark=0xd00 dev=eth0
		     -> xfrm state lookup matches
		     -> decrypted packet mark=0xd00
                        rule for 0x0d00 uses tbl 200
		     -> route decrypted packet to eth0
 		     -> bpf_packet runs clears mark and sends to stack

At this point a decrypted packet is sent to stack just as if it had arrived
decrypted. We could remove the steps after the xfrm decrypt if the xfrm stack
cleared the mark with an output_mark action in xfrm. We haven't done this yet
because it doesn't work pre-4.19 kernels. For ENI + overlapping subnets we
will put the 4.19 kernel requirement and remove the 0xd00 rule completely.
This simplifies datapath to just,

  eth0 (rx esp) -> bpf_packet mark -> decrypt -> stack ...

Much cleaner, but only usable where the output mark can be set. So we
keep the fallback path to use the mark if the mark action is not
available.

Before this patch two ip rules will be set for encrypt and decrypt,

 $ ip rule
   1:      from all fwmark 0xe00/0xf00 lookup 200
   1:      from all fwmark 0xd00/0xf00 lookup 200
   ...

After this patch only the encrypt rule will be set,

 $ ip rule
   1:      from all fwmark 0xe00/0xf00 lookup 200
   ...

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, remove mark from fwd rule

Make the fwd policy rule more permissive and remove the mark. This
way it will match in both cases: the mark is set because we are
running without subnet set, the case where the mark is 0 in the
ENI case.

Before this patch an example fwd policy was,

src 0.0.0.0/32 dst 192.168.32.0/19
        dir fwd priority 0 ptype main
        mark 0xd00/0xf00
        tmpl src 0.0.0.0 dst 192.168.62.241
                proto esp reqid 1 mode tunnel
                level use

Notice the 'mark 0xd00/0xf00' case restricting the match. After thiis
patch we drop the mark, which will make the match more permissive.

src 0.0.0.0/32 dst 192.168.32.0/19
        dir fwd priority 0 ptype main
        tmpl src 0.0.0.0 dst 192.168.62.241
                proto esp reqid 1 mode tunnel
                level use

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, clear mark from xfrm state in endpoint route modes

In endpoint route modes we have routes in the system to the endpoints so we
can skip the route back to the interface with bpf_network.o set, previously
used to do the redirect needed when the stack can not route.

This will work back to 4.14 kernels where XFRM_OUTPUT_MARK was added. If we
need to go back further an alternative route/mark/rule will need to be
added.

Before this patch we had in state rules like this,

src 0.0.0.0 dst 192.168.62.241
        proto esp spi 0x00000003 reqid 1 mode tunnel
        replay-window 0
        mark 0xd00/0xf00 output-mark 0xd00/0xf00

The above will use the 'output-mark' field to set the skb->mark after
decryption to 0xd00. But, after patch (and with endpoint routes enables)
we set the output-mark to 0 with the following state rule.

src 0.0.0.0 dst 192.168.62.241
        proto esp spi 0x00000003 reqid 1 mode tunnel
        replay-window 0
        mark 0xd00/0xf00 output-mark 0x0/0xf00

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, do not try to auto-discover if user provides iface

We currently ignore the user provided interface in ENI modes. Lets not
overwrite user input if they specify it.

Fixes: 2c5e192676 ("encryption/loader: Attach bpf_network program to correct interface for ENI mode")
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -68,7 +68,11 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 		return CTX_ACT_OK;
 	}
 	ctx->mark = 0;
+#if 0 //ENABLE_HOST_REDIRECT? What is the correct ifdef here
```
AFAIK, you want:
```suggestion
#ifdef ENABLE_ROUTING
```

`!defined(ENABLE_ROUTING)` <=> `enable-endpoint-routing=true`

jrfastab replied:
Thanks! That is what we want here.

jrfastab committed:
cilium: vendor, use cilium/netlink until upstream patches land

For encryption use cases we need the optional parameter in Xfrm
template. While we wait for those patches to land upstream point
at cilium/netlink with included patches.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -458,6 +458,26 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
+	if err != nil {
+		return nil, err
+	}
+	return addr[0].IPNet, nil
```
Is this deterministic?

joestringer replied:
For extra points: How does IPv4/IPv6 factor into this IP selection logic?

jrfastab replied:
Using IPv6 tunnel addresses has been carried around with limited testing and as far as I know no users. If we want IPv6 tunnels we probably need a feature request and more testing.

Why addr[0] is the correct address.  The IP is assigned to the interface and _should_ be routable. If user assigns a non-routable IP to the interface, then we have a problem yes. IMO we want a `cilium encryption status` command that dumps all the relevant info into a usable format so we can get visibility here and if it fails we will need user to somehow tell us what to do. I'm not sure we can otherwise automatically pick IPs here. With that in mind addr[0] seems the most logical choice to me.

joestringer replied:
Discussed out-of-band, IPv6 underlay for encryption likely needs some dedicated effort. Then for IPv4 it's relatively uncommon to have multiple addresses and in particular multiple addresses that would have different connectivity such that selecting the wrong one would cause connectivity issues.

OK to pass on this for now, as part of #14638 we can look at exposing visibility into the local IP selection to assist debugging of any potential issues in this area.

joestringer commented on bpf/lib/encrypt.h:
```diff
@@ -68,7 +68,11 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	return CTX_ACT_OK;
+#else
 	return redirect(CILIUM_IFINDEX, 0);
+#endif /* ENABLE_ROUTING */
```
This might work but it seems in conflict with the documented behaviour of the routing flag in userspace:

```
$ git grep -A 2 RequireRouting pkg/datapath/
pkg/datapath/config.go: // RequireRouting returns true if the endpoint requires BPF routing to
pkg/datapath/config.go- // be enabled, when disabled, routing is delegated to Linux routing
pkg/datapath/config.go: RequireRouting() bool
pkg/datapath/config.go-
pkg/datapath/config.go- // RequireEndpointRoute returns true if the endpoint wishes to have a
--
pkg/datapath/linux/config/config.go:    if e.RequireRouting() {
pkg/datapath/linux/config/config.go-            fmt.Fprintf(fw, "#define ENABLE_ROUTING 1\n")
pkg/datapath/linux/config/config.go-    }
```

jrfastab replied:
Agree its a bit of a conflict and might be fragile/broke. I guess the right thing to do is push RequireEndpointRoute into a define and use that.

jrfastab replied:
Updated to use a new endpoint routes enabled define.

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -159,9 +159,11 @@ func _ipSecReplacePolicyInFwd(src, dst, tmplSrc, tmplDst *net.IPNet, dir netlink
-	policy.Mark = &netlink.XfrmMark{
-		Value: linux_defaults.RouteMarkDecrypt,
-		Mask:  linux_defaults.IPsecMarkMaskIn,
+	if dir == netlink.XFRM_DIR_IN {
+		policy.Mark = &netlink.XfrmMark{
+			Value: linux_defaults.RouteMarkDecrypt,
+			Mask:  linux_defaults.IPsecMarkMaskIn,
+		}
```
nit: Would be nice to have some `ip x ...` output in these commit messages to see how the policies are different in each case, so we have something to easily refer to without having to stand up a cluster.

jrfastab replied:
updated commits with examples

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -523,7 +532,7 @@ func (n *linuxNodeHandler) enableSubnetIPsec(v4CIDR, v6CIDR []*net.IPNet) {
 				upsertIPsecLog(err, "getV6LinkLocalIP failed", ipsecIPv6Wildcard, cidr, spi)
 			}
-			spi, err := ipsec.UpsertIPsecEndpoint(linkAddr, ipsecIPv6Wildcard, cidr, ipsec.IPSecDirIn)
+			spi, err := ipsec.UpsertIPsecEndpoint(linkAddr, ipsecIPv6Wildcard, cidr, ipsec.IPSecDirIn, false)
```
I'm not sure I understand the significance of passing `false` for this and subsequent changes in the rest of this file vs. earlier. What's the intent here?

jrfastab replied:
This will determine if the packet is routed back to the ingress interface or not. If true the packet will be routed directly from xfrm decrypt operation with being sent to program running on ingress interface. If false we will continue to run through ingress interface decrypted.

We could also set enableIPsec case true as well, but I didn't here and will to that in a follow up as its getting close to a feature at that point.

joestringer replied:
I see, I think the piece of context I was missing is that this may only be true for subnet mode + endpoint routes (just like the commit message for that change implies).

joestringer reviewed (changes requested):
Would be good to at least update the following docs with the new requirements:
* https://docs.cilium.io/en/stable/operations/system_requirements/#linux-kernel 
* https://docs.cilium.io/en/stable/operations/system_requirements/#required-kernel-versions-for-advanced-features

I wonder if we could probe the new feature dependency, eg by creating a namespace on startup & trying to send the kernel a message including `XFRM_OUTPUT_MARK` to see whether it gives us a `EINVAL` or similar..?

A few minor nits / questions below, determinism of arbitrarily picking the first IP assigned to an interface seems like the most concerning.

joestringer reviewed (approved):
With the extra `ip x ...` output this is easier to work through and at least cross-reference the expected effect of the changes (especially for someone like me with not much kernel ipsec background). LGTM.

I still feel like it'd be nice to drop a note somewhere about the limitations but I'll concede that it's pretty unlikely that someone will try to deploy a v4.9.x kernel with ipsec these days, particularly in these managed clouds which tend to provide standard distros with newer kernels. At least this issue should be searchable if anyone tries it out in future and it'll be clear there's a limitation.

aditighag commented on pkg/datapath/loader/base.go:
```diff
@@ -357,13 +357,15 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
-			iface, err := linuxrouting.RetrieveIfaceNameFromMAC(mac.String())
-			if err != nil {
-				log.WithError(err).WithField("mac", mac).Fatal("Failed to set encrypt interface in the ENI ipam mode")
+		if option.Config.EncryptInterface == "" {
```
How does this work if user specifies a different interface from the one `cilium_host` is associated to? Encrypted traffic will arrive at the destination node on the primary ENI interface, but if the `bpf_network` program to decrypt traffic is attached to a different interface selected by users, then packets won't get decrypted? Are we expecting users to provide the correct interface by looking at the `cilium_host` ENI association?

joestringer replied:
~In this series with the subnet mode, commit `cilium: encryption, ENI with subnet encryption needs rules for link IP` (currently bc534be5e29343f4bd34a7aaa7b883df37ecf8ed) is proposing to switch away from reliably using the cilium_host IP for the local encryption IP to instead pick an IP from the `option.Config.EncryptInterface`. So there won't be a dependency on the cilium_host IP any more for this configuration.~

EDIT: Actually, on second thought the above commit is all about the "in" rules so it doesn't make sense.

joestringer replied:
I've got the impression a few times from discussions with John that in this mode, the IP used for encryption isn't necessarily the cilium_host IP but I don't really understand where it comes from. @jrfastab is that true? Is there something about subnet mode that causes different IPs to be used for the encryption?

That said, this particular patch is basically just saying "if the user explicitly configures the encrypt interface, do what they say, even if it's wrong because if they're explicitly setting this configuration, they know what they're doing". In general we should not automatically override some explicit user configuration.

aditighag replied:
>Are we expecting users to provide the correct interface by looking at the cilium_host ENI association?

>In general we should not automatically override some explicit user configuration.

I have a docs PR in progress where I mention that the `option.Config.EncryptInterface` config won't take effect in case of the IPAM ENI mode.

jrfastab replied:
Problem is at least in EKS (k8s 1.19) the detected interface is wrong, eth1 instead of eth0 in my case.  I haven't investigated yet why its the wrong interface yet (assumption below), but merely noting we shouldn't ignore user input. If a user pushes down an explicit interface then we should honor it. If we want a WARN or something if they don't match that would be nice I guess.

In subnet mode, required in modes with a global IP pool, we set src IP using the IPV4_ENCRYPT_IFACE chosen here,
```
        if iface := option.Config.EncryptInterface; iface != "" {
                link, err := netlink.LinkByName(iface)
                if err == nil {
                        cDefinesMap["ENCRYPT_IFACE"] = fmt.Sprintf("%d", link.Attrs().Index)

                        addr, err := netlink.AddrList(link, netlink.FAMILY_V4)
                        if err != nil {
                                return err
                        }
                        if len(addr) == 0 {
                                return fmt.Errorf("no IPv4 addresses available in encrypt interface %q", iface)
                        }
                        a := byteorder.HostSliceToNetwork(addr[0].IPNet.IP, reflect.Uint32).(uint32)
                        cDefinesMap["IPV4_ENCRYPT_IFACE"] = fmt.Sprintf("%d", a)
                }
        }
```
Why not cilum_host in this mode? The initial idea here was to avoid leaking podIPs and CiliumIPs into the network and only use networkIPs. I assume this is why the auto detection is failing, some mismatch between above and the selection logic earlier. Anyways when given multiple network facing interfaces we should attach to all interfaces. Its probably fine to assume that if we send traffic on an interface it will be returned on that interface, but thats an assumption that we could get rid of by attaching to all interfaces.

joestringer replied:
Huh, weird, so that logic you point to above doesn't have any link to the logic that this comment thread is looking at. Seems like if you don't configure `option.Config.EncryptInterface` then `ENCRYPT_IFACE` and  `IPV4_ENCRYPT_IFACE` won't get set. Might explain why the autodetection isn't lining up with the other parts of the datapath for your setup.

aditighag replied:
>In subnet mode, required in modes with a global IP pool, we set src IP using the IPV4_ENCRYPT_IFACE chosen here,

To keep the scope of the original fix for the IPAN ENI issue limited, I left this as a follow-up with a note in the original PR #https://github.com/cilium/cilium/pull/14924. :) In the code snippet you've added above, it also mandates that the interface an an IP address. That's not the case for the IPAN ENI mode.

jrfastab replied:
Agree I think it needs a follow up. What do you mean by 'thats not the case for the IPAM ENI mode"? Do you mean in AWS-ENI? My eth0 has an IP address in that configuration,
```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
    link/ether 02:52:43:a6:f3:43 brd ff:ff:ff:ff:ff:ff
    inet 192.168.90.141/19 brd 192.168.95.255 scope global dynamic eth0
       valid_lft 3056sec preferred_lft 3056sec
    inet6 fe80::52:43ff:fea6:f343/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
    link/ether 02:67:c8:f5:87:87 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::67:c8ff:fef5:8787/64 scope link 
       valid_lft forever preferred_lft forever
```
eth1 however does not.

aditighag replied:
Yes, that's exactly what I was referring to. On EKS, eth1 will be auto-detected.

aditighag replied:
Logging a warning sounds like a decent middle ground.

aditighag reviewed (approved):
LGTM, thanks!
